core/src/main/java/com/linecorp/armeria/common/stream/TimeoutStreamMessage.java

```diff
@@ -0,0 +1,273 @@
+     */
+    private void cancelSchedule() {
+        if (timeoutSubscriber != null) {
+            timeoutSubscriber.cancelSchedule();
+        }
+    }
+
+    /**
+     * Aborts the stream with an {@link AbortedStreamException} and prevents further subscriptions.
+     * Calling this method on a closed or aborted stream has no effect.
+     * Cancels the timeout schedule if it is set.
+     */
+    @Override
+    public void abort() {
+        cancelSchedule();
+        delegate.abort();
+    }
+
+    /**
+     * Aborts the stream with the specified {@link Throwable} and prevents further subscriptions.
+     * Calling this method on a closed or aborted stream has no effect.
+     * Cancels the timeout schedule if it is set.
+     *
+     * @param cause the cause of the abort
+     */
+    @Override
+    public void abort(Throwable cause) {
+        cancelSchedule();
+        delegate.abort(cause);
+    }
+
+    static final class TimeoutSubscriber<T> implements Runnable, Subscriber<T>, Subscription {
+
+        private static final String TIMEOUT_MESSAGE = "Stream timed out after %d ms (timeout mode: %s)";
+        private final Subscriber<? super T> delegate;
+        private final EventExecutor executor;
+        private final StreamTimeoutMode timeoutMode;
+        private final Duration timeoutDuration;
+        private final long timeoutNanos;
+        @Nullable
+        private ScheduledFuture<?> timeoutFuture;
+        @Nullable
+        private Subscription subscription;
+        private long lastEventTimeNanos;
+        private boolean completed;
+        private boolean canceled;
+
+        TimeoutSubscriber(Subscriber<? super T> delegate, EventExecutor executor, Duration timeoutDuration,
+                          StreamTimeoutMode timeoutMode) {
+            this.delegate = requireNonNull(delegate, "delegate");
+            this.executor = requireNonNull(executor, "executor");
+            this.timeoutDuration = requireNonNull(timeoutDuration, "timeoutDuration");
+            timeoutNanos = timeoutDuration.toNanos();
+            this.timeoutMode = requireNonNull(timeoutMode, "timeoutMode");
+        }
+
+        private ScheduledFuture<?> scheduleTimeout(long delay, TimeUnit unit) {
+            return executor.schedule(this, delay, unit);
+        }
+
+        void cancelSchedule() {
+            if (timeoutFuture != null && !timeoutFuture.isCancelled()) {
+                timeoutFuture.cancel(false);
+            }
+        }
+
+        @Override
+        public void run() {
+            if (timeoutMode == StreamTimeoutMode.UNTIL_NEXT) {
+                final long currentTimeNanos = System.nanoTime();
+                final long elapsedNanos = currentTimeNanos - lastEventTimeNanos;
+
+                if (elapsedNanos < timeoutNanos) {
+                    final long delayNanos = timeoutNanos - elapsedNanos;
+                    timeoutFuture = scheduleTimeout(delayNanos, TimeUnit.NANOSECONDS);
```

Comment:
We always use the nanoseconds time unit so it can be inlined.
`executor.schedule(this, delay, TimeUnit.NANOSECONDS)`
```suggestion
                    timeoutFuture = scheduleTimeout(delayNanos);
```